dotnet/samples/KernelSyntaxExamples/Example88_CustomMessageAPIEndpoint.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Connectors.OpenAI;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Examples;
+
+/// <summary>
+/// This example shows a way of using OpenAI connector with other APIs that supports the same ChatCompletion Message API standard from OpenAI.
+///
+/// To proceed with this example will be necessary to follow those steps:
+/// 1. Install LMStudio Platform in your environment
+/// 2. Open LM Studio
+/// 3. Search and Download both Phi2 and Llama2 models (preferably the ones that uses 8GB RAM or more)
+/// 4. Start the Message API Server on http://localhost:1234
+/// 5. Run the examples.
+///
+/// OR
+///
+/// 1. Start the Ollama Message API Server on http://localhost:11434 using docker
+/// 2. docker run -d --gpus=all -v "d:\temp\ollama:/root/.ollama" -p 11434:11434 --name ollama ollama/ollama <see href="https://ollama.com/blog/ollama-is-now-available-as-an-official-docker-image" />
+/// 3. Set Llama2 as the current ollama model: docker exec -it ollama ollama run llama2
+/// 4. Run the Ollama examples.
+///
+/// OR
+///
+/// 1. Start the LocalAI Message API Server on http://localhost:8080
+/// 2. docker run -ti -p 8080:8080 localai/localai:v2.12.3-ffmpeg-core phi-2 <see href="https://localai.io/docs/getting-started/run-other-models/" />
+/// 3. Run the LocalAI examples.
+/// </summary>
+public class Example88_CustomMessageAPIEndpoint : BaseTest
+{
+    [Theory(Skip = "Manual configuration needed")]
+    [InlineData("LMStudio", "http://localhost:1234", "llama2")] // Setup Llama2 as the model in LM Studio UI and start the Message API Server on http://localhost:1234
+    [InlineData("Ollama", "http://localhost:11434", "llama2")] // Start the Ollama Message API Server on http://localhost:11434 using docker
+    [InlineData("LocalAI", "http://localhost:8080", "phi-2")]
+    public async Task LocalModel_ExampleAsync(string messageAPIPlatform, string url, string modelId)
+    {
+        WriteLine($"Example using local {messageAPIPlatform}");
+        // Setup Llama2 as the model in LM Studio UI.
+
+        var kernel = Kernel.CreateBuilder()
+            .AddOpenAIChatCompletion(
+                modelId: modelId,
+                apiKey: null,
+                endpoint: new Uri(url))
+            .Build();
```

Comment:
I wonder whether this solution was considered:
```csharp
using var httpClient = new HttpClient();
httpClient.BaseAddress = new Uri(url);

var kernel = Kernel.CreateBuilder()
    .AddOpenAIChatCompletion(
        modelId: modelId,
        apiKey: "an api key",
        httpClient: httpClient)
    .Build();
```

Reply:
I don't think that's viable:
1) HttpClient.BaseAddress is only used when the supplied address is relative. If the Uri is absolute, the BaseAddress is ignored.
2) We can't modify the BaseAddress of an HttpClient that's provided to us, only if we completely own it, both because we could be changing the BaseAddress in a way someone didn't expect, and because setting BaseAddress will fail if the client has already been used to make requests.